src/contexts/ethereum-provider.context.tsx

```diff
@@ -0,0 +1,119 @@
+import { createContext, FC, useContext, useEffect, useState } from "react";
+import { Web3Provider } from "@ethersproject/providers";
+import WalletConnectProvider from "@walletconnect/ethereum-provider";
+import { useNavigate } from "react-router-dom";
+
+import { EthereumEvent, MetaMaskProvider, WalletName } from "src/domain";
+import { AsyncTask } from "src/utils/types";
+import { ethereumAccountsParser } from "src/adapters/parsers";
+import { getConnectedAccounts } from "src/adapters/ethereum";
+import { useEnvContext } from "./env.context";
+import routes from "src/routes";
+
+interface EthereumProviderContextData {
+  provider?: Web3Provider;
+  account: AsyncTask<string, string>;
+  connectProvider?: (walletName: WalletName) => Promise<void>;
+}
+
+const EthereumProviderContext = createContext<EthereumProviderContextData>({
+  provider: undefined,
+  account: { status: "pending" },
+});
+
+const EthereumProviderProvider: FC = (props) => {
+  const navigate = useNavigate();
+  const [rawProvider, setRawProvider] = useState<MetaMaskProvider | WalletConnectProvider>();
+  const [provider, setProvider] = useState<Web3Provider>();
+  const [account, setAccount] = useState<AsyncTask<string, string>>({ status: "pending" });
+  const env = useEnvContext();
+
+  const connectProvider = async (walletName: WalletName): Promise<void> => {
+    switch (walletName) {
+      case WalletName.METAMASK: {
+        if (window.ethereum) {
```

Comment:
I think we should check the `ExternalProvider` to guarantee that it actually comes from Metamask. Like it's now we will accept a provider different from Metamask present on `window.ethereum`.

Regarding how to check it, at least, we should check for the prop `isMetaMask`.

A more ambitious approach would be to actually decode the object, which would let us declare the type `ethereum` from [`react-app-env.d.ts`](#diff-433083e9966b3a808a0b2dce3e47b0b74152c61cb7d5b6a4ad8aad2592dacd81R7) as `unknown` (what it actually is) until it's properly decoded and added to the context.

`ExternalProvider` is indeed a pretty simple interface and [zod allows us to parse objects with functions](https://github.com/colinhacks/zod#function-schemas).

Reply:
I've given this a try and works beautifully.

In `src/react-app-env.d.ts` we set `ethereum` as `unknown`:

```typescript
/// <reference types="react-scripts" />

declare global {
  interface Window {
    ethereum: unknown; // at this point is all we know
  }
}

export {};
```

Later in `src/adapters/parsers.ts` we declare the parser of the provider, making sure that it's a valid **MetaMask** `ExternalProvider` (notice the `isMetaMask: z.boolean().refine((val) => val === true)`):

```typescript
const metamaskExternalProviderParser = StrictSchema<ExternalProvider>()(
  z
    .object({
      isMetaMask: z.boolean().refine((val) => val === true),
      isStatus: z.boolean(),
      host: z.string(),
      path: z.string(),
      sendAsync: z
        .function()
        .args(
          z.object({
            method: z.string(),
            params: z.array(z.any()).optional(),
          }),
          z.function().args(z.any(), z.any()).returns(z.void())
        )
        .returns(z.void()),
      send: z
        .function()
        .args(
          z.object({
            method: z.string(),
            params: z.array(z.any()).optional(),
          }),
          z.function().args(z.any(), z.any()).returns(z.void())
        )
        .returns(z.void()),
      request: z
        .function()
        .args(
          z.object({
            method: z.string(),
            params: z.array(z.any()).optional(),
          })
        )
        .returns(z.promise(z.void())),
    })
    .partial()
);
```

And finally in `src/contexts/ethereum-provider.context.tsx`, we decode the object and only proceed when it's a valid **MetaMask** `ExternalProvider`:

```typescript
const parsedMetamaskExternalProvider = metamaskExternalProviderParser.safeParse(window.ethereum);
if (parsedMetamaskExternalProvider.success) {...}
```

I experimented with this parser just for fun. I realize that going this way has the drawback of having to maintain the parser whenever the interface `ExternalProvider` is changed, so I'm actually not sure if this is the way to go.

At least though, we should tests for `window.ethereum && window.ethereum.isMetaMask` in the `if` statement.

Regarding this, I've confirmed that with the current implementation you can log in to the bridge using the [Coinbase Wallet extension](https://chrome.google.com/webstore/detail/coinbase-wallet-extension/hnfanknocfeofbddgcijnmhnfnkdnaad), an this may have undesired consequences.